pipeline/runtime/runtestsV2.go

```diff
@@ -0,0 +1,330 @@
+}
+
+func collectTestReportsAndCg(ctx context.Context, log *logrus.Logger, r *api.StartStepRequest, start time.Time, stepName string, tiConfig *tiCfg.Cfg) error {
+	cgStart := time.Now()
+
+	cgErr := callgraph.Upload(ctx, stepName, time.Since(start).Milliseconds(), log, cgStart, tiConfig, outDir)
+	if cgErr != nil {
+		log.WithField("error", cgErr).Errorln(fmt.Sprintf("Unable to collect callgraph. Time taken: %s", time.Since(cgStart)))
+		cgErr = fmt.Errorf("failed to collect callgraph: %s", cgErr)
+	}
+
+	reportStart := time.Now()
+	crErr := report.ParseAndUploadTests(ctx, r.TestReport, r.WorkingDir, stepName, log, reportStart, tiConfig, r.Envs)
+	if crErr != nil {
+		log.WithField("error", crErr).Errorln(fmt.Sprintf("Failed to upload report. Time taken: %s", time.Since(reportStart)))
+	}
+	return cgErr
+}
+
+// Second parameter in return type (bool) is will be used to decide whether the filter file should be created or not.
+// In case of running all the cases no filter file should be created.
+func getTestsSelection(ctx context.Context, fs filesystem.FileSystem, stepID, workspace string, log *logrus.Logger,
+	isManual bool, tiConfig *tiCfg.Cfg) (types.SelectTestsResp, bool) {
+
+	selection := types.SelectTestsResp{}
+	var RunOnlySelectedTests = true // Enabled true by default
+
+	if isManual {
+		log.Infoln("Manual execution has been detected. Running all the tests")
+		RunOnlySelectedTests = false
+		return selection, false
+	}
+
+	var files []types.File
+	var err error
+
+	if utils.IsPushTriggerExecution(tiConfig) {
+		lastSuccessfulCommitID, commitErr := utils.GetCommitInfo(ctx, stepID, tiConfig)
+		if commitErr != nil {
+			log.Infoln("Failed to get reference commit", "error", commitErr)
+			RunOnlySelectedTests = false // TI selected all the tests to be run
+			return selection, false
+		}
+
+		if lastSuccessfulCommitID == "" {
+			log.Infoln("Test Intelligence determined to run all the tests to bootstrap")
+			RunOnlySelectedTests = false // TI selected all the tests to be run
+			return selection, false
+		}
+
+		log.Infoln("Using reference commit: ", lastSuccessfulCommitID)
+		files, err = utils.GetChangedFilesPush(ctx, workspace, lastSuccessfulCommitID, tiConfig.GetSha(), log)
+		if err != nil {
+			log.Errorln("Unable to get changed files list. Running all the tests.", "error", err)
+			RunOnlySelectedTests = false
+			return selection, false
+		}
+	} else {
+		files, err = utils.GetChangedFilesPR(ctx, workspace, log)
+		if err != nil || len(files) == 0 {
+			log.Errorln("Unable to get changed files list for PR. Running all the tests.", "error", err)
+			RunOnlySelectedTests = false
+			return selection, false
+		}
+	}
+
+	filesWithpkg := java.ReadPkgs(log, fs, workspace, files)
+	selection, err = utils.SelectTests(ctx, workspace, filesWithpkg, RunOnlySelectedTests, stepID, fs, tiConfig)
+	if err != nil {
+		log.WithError(err).Errorln("There was some issue in trying to figure out tests to run. Running all the tests")
```

Comment:
`An unexpected error occurred during test selection. Running all tests.`

Reply:
How do we debug the failures? Should we have a env flag such as `TI_DEBUG` for printing debug logs (including error codes/strings)?